### ISSUE TITLE
PlaceBeside and Chunk Boundaries

### ISSUE THREAD

**Issue by Draco18s** [lawremi/CustomOreGen#78]
It does not appear that the PlaceBeside descriptor works properly across chunk boundaries.  I'm having a lot of trouble getting a distribution to work and noticed the chunk edges as the only common factor for some failed placements.  This is even when using a deferred radius of 32.  I suspect that this is because of chunk generation order, the first chunk to populate won't see the adjacency blocks of the second chunk to populate (when the block to look for is a block placed by the another distribution).

For Place above/below there are issues as well, but I haven't been able to figure out why.  When used by themselves, they work perfectly.  When used in a separate cloud, but before or after a placebeside, they outright fail.  Placed after, the above fails (and below is spotty).  Placed before, the sides fail.

Here's a distribution (you'll need the rest of the config stuff, but thats easy, and then I don't have to send you the whole file with 8 different distributions in it), its already set up to use wool and glass and places in the air:

http://pastebin.com/qGjd1yw9
Seed: -2627735460604631124
X: -250
Z: -210

Chunk edge errors:
![2015-08-11_11 13 19](https://cloud.githubusercontent.com/assets/1750680/9201412/01c3f332-401b-11e5-9adf-8afd13bb4ec1.png)
![2015-08-11_11 12 03](https://cloud.githubusercontent.com/assets/1750680/9201443/281a185e-401b-11e5-842b-ab1c38513e1d.png)
Missing tops:
![2015-08-11_11 13 37](https://cloud.githubusercontent.com/assets/1750680/9201413/01c443d2-401b-11e5-9ef9-60ab82ff22a3.png)
Inconsistent bottoms:
![2015-08-11_11 16 39](https://cloud.githubusercontent.com/assets/1750680/9201414/01c73330-401b-11e5-805d-1f95e134f190.png)
I just noticed that actually the expansion on the *sides* is messed up there, too.  Some of the green wool wasn't expanded horizontally to orange either, not just downwards.

If you move the clouds marked "Sides" to the top of the distribution, you'll see almost no side adjacent blocks placed what so ever, but all of the above/below work perfectly (albeit, missing the orange for some reason).
![2015-08-11_11 30 40](https://cloud.githubusercontent.com/assets/1750680/9201712/66d8904c-401c-11e5-8e8e-06fab90beb76.png)

Anyway, I've tried organizing this distribution a dozen different ways and this is the *closest* I can come to something that works correctly.  I've tried parent/child distributions, I've tried sibling (making no parent/child clouds gets the *worst* result with almost no adjacency success at all), I've tried multiple top-level distributions.

**Comment by Draco18s** [comment]
Examined this a a little farther.  In the base-case-scenario (Above, Below, Sides) it looks like each pass only manages to place 1 block, its child distribution then fails, not seeing the parent's blocks.

I.e. above runs first and places the brown wool, but no green (each green attempt sees only air).
below runs next and places down green (seeing the brown from the above pass) but none of its own orange.  The glass, which is a sibling, does see the placed green.
sides runs last, placing its green adjacent to the brown from the first pass, orange adjacent to the green from the second pass (but not its own), and then the glass (as a sibling).

Ah, yes.  That's what's happening alright.  Child distributions aren't seeing their parent's blocks for adjacency.

**Comment by lawremi** [comment]
Not sure what we can do about the chunk boundary. There are all sorts of order issues that come into play with these adjacency rules. One could imagine complex deferral schemes, but the recursion makes my mind explode.

**Comment by Draco18s** [comment]
Yeah...Wish I had realized the child/parent ordering sooner and then I'd have said "meh" to the chunk boundary problem and moved on with my life.

And whatever I did to the distribution recently has taken a huge toll on my world generation speed (< 1% a minute).  Game isn't even utilizing the full amount of CPU it normally does during world gen (normally it'll try and use as much as it can, about 35%, right now it's taking 17%).

Doesn't help that the game isn't pulling the new XML files either, but "helpfully" caching them. 9..9
